src/pages/index.tsx

```diff
@@ -3,20 +3,157 @@ import { type NextPage } from "next";
 import { Layout } from "../components/Layout";
 import { useRouter } from "next/router";
 import dynamic from "next/dynamic";
-import { Posts } from "../components/Post";
+import { PostList } from "../components/PostList";
+import { api } from "../utils/api";
+import { useState } from "react";
+import { useDebounce } from "../hooks/useDebounce";
+import { Pagination } from "../components/Pagination";
+import { PostListFiltersValues, SearchPost } from "../components/SearchPost";
+import { z } from "zod";
+import { BlogPostCategory } from "../types/blogPost";
+import qs from "qs";
+
+const POST_LIST_FILTERS_DEBOUNCE = 500;
+const POSTS_PER_PAGE = 8;
+
+const queryParamsSchema = z.object({
+  search: z.string().optional().default(""),
+  categories: z.array(z.nativeEnum(BlogPostCategory)).optional().default([]),
+  activePage: z.coerce.number().default(0),
+});
+
+const parseQueryParams = (
+  rawQueryParams: RawQueryParams
+): z.infer<typeof queryParamsSchema> => {
+  const parsedQueryParams = queryParamsSchema.safeParse({
+    categories: rawQueryParams.categories?.split(","),
+    search: rawQueryParams.search,
+    activePage: rawQueryParams.activePage,
+  });
+
+  if (parsedQueryParams.success) {
+    return parsedQueryParams.data;
+  }
+
+  return {
+    activePage: 0,
+    categories: [],
+    search: "",
+  };
+};
+
+interface RawQueryParams {
+  search?: string;
+  categories?: string;
+  activePage?: string;
+}
 
 const HomePage: NextPage = () => {
+  const rawQueryParams = qs.parse(window.location.search, {
+    ignoreQueryPrefix: true,
+  }) as RawQueryParams;
+
+  const parsedQueryParams = parseQueryParams(rawQueryParams);
+
+  const [activePage, setActivePage] = useState(parsedQueryParams.activePage);
+  const [postListFilters, setPostListFilters] = useState<PostListFiltersValues>(
+    {
+      categories: parsedQueryParams.categories,
+      search: parsedQueryParams.search,
+    }
+  );
   const router = useRouter();
+
+  const deboucedPostListFilters = useDebounce(
+    postListFilters,
+    POST_LIST_FILTERS_DEBOUNCE
+  );
+
+  const blogPostQuery = api.blogPost.getBlogPosts.useQuery({
+    page: activePage,
+    perPage: POSTS_PER_PAGE,
+    categories: postListFilters.categories,
+    search: deboucedPostListFilters.search,
+  });
+
+  const handlePostListFiltersChange = async (
+    filters: PostListFiltersValues
+  ) => {
+    setPostListFilters(filters);
+    await router.push(
+      {
+        query: qs.stringify(
+          { ...filters, activePage },
+          {
+            arrayFormat: "comma",
+          }
+        ),
+      },
+
+      undefined,
+      {
+        shallow: true,
+      }
+    );
+    setActivePage(0);
+  };
+
+  const handleUrlAll = async (activePage: number) => {
```

Comment:
`handle` prefix for functions is usually used when you want to handle some event. Otherwise you should name function based on it's responsibility. In this case for example `updateQueryParams`.

I think this function should take `(filters:PostListFiltersValues, activePage: number)` as parameters. And then, you could easily use this function on both filters and page change